main.go

```diff
@@ -43,26 +45,127 @@ func main() {
 	}
 
 	// basic metrics + client metrics + exporter own metrics (ProcessCollector and GoCollector)
+	// NOTE: This handler was retained for backward compatibility. See: https://jira.percona.com/browse/PMM-1901.
 	{
-		prometheus.MustRegister(basic.New(cfg, sess))
-		prometheus.MustRegister(client)
-		http.Handle(*basicMetricsPathF, promhttp.HandlerFor(prometheus.DefaultGatherer, promhttp.HandlerOpts{
+		basicCollector := basic.New(cfg, sess)
+
+		registry := prometheus.NewRegistry()
+		registry.MustRegister(prometheus.NewProcessCollector(os.Getpid(), "")) // from prometheus.DefaultGatherer
+		registry.MustRegister(prometheus.NewGoCollector())                     // from prometheus.DefaultGatherer
+
+		registry.MustRegister(basicCollector)
+		registry.MustRegister(client)
+		http.Handle(*basicMetricsPathF, promhttp.HandlerFor(registry, promhttp.HandlerOpts{
 			ErrorLog:      log.NewErrorLogger(),
 			ErrorHandling: promhttp.ContinueOnError,
 		}))
 	}
 
+	// This collector should be only one for both cases, because it creates goroutines which sends API requests to Amazon in background.
+	enhancedCollector := enhanced.NewCollector(sess)
+
 	// enhanced metrics
+	// NOTE: This handler was retained for backward compatibility. See: https://jira.percona.com/browse/PMM-1901.
 	{
 		registry := prometheus.NewRegistry()
-		registry.MustRegister(enhanced.NewCollector(sess))
+		registry.MustRegister(enhancedCollector)
 		http.Handle(*enhancedMetricsPathF, promhttp.HandlerFor(registry, promhttp.HandlerOpts{
 			ErrorLog:      log.NewErrorLogger(),
 			ErrorHandling: promhttp.ContinueOnError,
 		}))
 	}
 
-	log.Infof("Basic metrics   : http://%s%s", *listenAddressF, *basicMetricsPathF)
-	log.Infof("Enhanced metrics: http://%s%s", *listenAddressF, *enhancedMetricsPathF)
+	// all metrics (with filtering)
+	{
+		// Create separate instance of basic collector and remove metrics which cross with enhanced collector.
+		basicCollector := basic.New(cfg, sess)
+		basicCollector.ExcludeMetrics("CPUUtilization", "FreeStorageSpace", "FreeableMemory")
```

Comment:
Is there any test to check that we don't exclude these metrics when we collect only basic?
Looks like we should use another endpoint to get only basic metrics. But we left that only for backward compatibility.
So my main point is that we can loose some data from basic metrics if we request only them.
Maybe it's better to handle it somewhere inside handler?

Reply:
It seems you are right. I'll check it, thanks.